docs/RFCs/19042021_Cert_Management_In_CRDB_Helm_Chart.md

```diff
@@ -0,0 +1,243 @@
+ 
+   | Resource          | Hook-weight   | Order of Installation     |
+   |----------------   |-------------  |-----------------------    |
+   | ServiceAccount    | 1             | 1st                       |
+   | Role              | 2             | 2nd                       |
+   | RoleBindings      | 3             | 3rd                       |
+   | Job               | 4             | 4th                       |
+ 
+ * After all the `pre-install` hooks completed successfully, they will be deleted by hook deletion-policy defined in
+ annotations.
+ 
+ * All the certificate generation-related info will be passed on to the `pre-install` job as env variables.
+    ```yaml
+    env:
+       - name: CA_CERT_DURATION
+         value: {{ default 3650 .Values.tls.certs.generate.caCertDuration}}
+       - name: NODE_CERT_DURATION
+         value: {{ default 365 .Values.tls.certs.generate.nodeCertDuration}}
+       - name: Client_CERT_DURATION
+         value: {{ default 365 .Values.tls.certs.generate.clientCertDuration}}
+       {{- if and (tls.certs.generate.caProvided .Values.tls.certs.generate.caSecret) }}
+           {{- if not (lookup "v1" "Secret" ".Release.Namespace" ".Values.tls.certs.generate.caSecret")}}
+           {{ fail "CA secret doesn't exist in cluster"}}
+           {{- end }}
+       - name: CA_SECRET
+         values:
+       {{- end }}
+    ```
+ 
+* 3 empty secret will be created in Helm chart for `cockroachdb-ca`, `cockroachdb-node` and `cockroachdb-root` if `tls.certs.generate.enabled`
+ is set.
+  * Data to these secrets will be populated in the `pre-install` job.
+  * In case CA is provided by the user, then `cockroachdb-ca` secret is skipped.
+  * Annotation is set on all the secrets created by CockroachDB; eg: `managed-by: cockroachdb`
+ 
+* A cronjob will be created in Helm chart when `tls.certs.generate.rotateCerts` is set.
+  * This cronjob will run periodically to rotate the certificates.
+  * The schedule of the cronjob will be of minimum `clientCertDuration` and `nodeCertDuration` minus some delta.
+  * On every scheduled run, it will check if there is any certificate that is going to expire before the next scheduled run,
+   if yes then it will renew the certificates.
+  
+  * <b>The cronjob will use the same `pre-install` job image for certificate rotations. The `pre-install` job image binary will have an argument `--rotate` for handling certificate rotation.</b>
+ 
+* The Stateful pod will be changed to only run `copy-certs` initContainer to copy the certificates from Node secret to emptyDir volume.
+  The rest of the main DB container flow will remain the same.
+* The post-install job will be changed to only run `copy-certs` initContainer to copy the certificates from Client secret to emptyDir volume.
+  The rest of the main cluster-init container flow will remain the same.
+ 
+`TODO: Need to identify how to generate the SIGHUP signal in all the Nodes for certificate renewal`
```

Comment:
One way that we've done this is to install a sidecar container in the CockroachDB pod that listens for changes to the certificate files and SIGHUPs the CRDB process if one is detected. We could introduce this functionality to our Helm chart.

Reply:
there are two patterns here, rotating node certs that share a CA, and rotating a CA.  The first is a rolling restart, the latter is a full (parallel) restart

Reply:
I would rather implement a rolling restart than a SIGHUP because managing sidecar containers would be a large increase in the scope of work here.

Reply:
I don't think a sidecar container is actually that much more work? We have the code for this in Cockroach Cloud, and we could easily open-source the few lines of code that it takes to do this. A SIGHUP is less disruptive to the cluster than restarting every node.

If either solution was the same amount of work, would you agree that a SIGHUP would be better?

Reply:
the problem is that we then have to distribute and maintain this additional sidecar container, including releasing new versions of it if there's a security patch, etc. It's the O&M overhead that I'm concerned about more than the actual work adding it to the helm chart itself. Does that make sense @DuskEagle ?

Reply:
I see your reasoning. I still think we should be providing a way to reload certificates when they change on disk, either via a sidecar container or by native functionality in CRDB. But I'm fine with tabling this discussion and leaving it as a separate item from this RFC.